### ISSUE TITLE
HAProxy returns incorrect (ending with tons of 0x00) content

### ISSUE THREAD

**Issue by trebonius0** [haproxy/haproxy#1510]
### Detailed Description of the Problem

Hello

I noticed today that haproxy returns incorrect data on some calls since last update (2.4.11)

I've clients calling my haproxy, which then calls my apache servers. My apache servers return some content, haproxy returns it, and the md5 of the two content (apache response vs haproxy response) is not the same
I checked the 2 contents, the apache content is correct and for some reason haproxy modifies it. 

In particular, the end of the file haproxy returns is full of 0x00 (when it isn't in the file apache returns)
The files returned by haproxy and apache2 have the same size though.

If I call haproxy multiple times requesting for the same content, the md5 will vary from time to time, so I guess the place in the content where all those 0x00 start appearing do change

### Expected Behavior

I expect the data returned by haproxy be the same as the data returned by the server it calls

### Steps to Reproduce the Behavior

Using the same config for haproxy as mine (in the next fields of the bug report), try to return the file contained in this zip
[apache6.zip](https://github.com/haproxy/haproxy/files/7838614/apache6.zip), using http

The content of the file will not be correct (here is what i got for instance) [response6.zip](https://github.com/haproxy/haproxy/files/7838623/response6.zip) (the file contained in the zip)
 

### Do you have any idea what may have caused this?

Latest update maybe be the cause, as I updated yesterday and users started to complain today. 
Nevertheless, it's a bit hard in my configuration to rollback to previous haproxy version

### What is your configuration?

```haproxy
global
    log /dev/log    local0
    log /dev/log    local1 notice
    chroot /var/lib/haproxy
    stats socket /run/haproxy/admin.sock mode 660 level admin expose-fd listeners
    stats timeout 30s
    user haproxy
    group haproxy
    daemon
    maxconn 320000
    tune.ssl.cachesize 1000000
    ca-base /etc/ssl/certs
    crt-base /etc/ssl/private
    ssl-default-bind-ciphers ECDH+AESGCM:DH+AESGCM:ECDH+AES256:DH+AES256:ECDH+AES128:DH+AES:RSA+AESGCM:RSA+AES:!aNULL:!MD5:!DSS
    ssl-default-bind-options no-sslv3

    lua-load /etc/haproxy/200.lua

defaults
    log    global
    mode    http
    option    httplog
    timeout connect 10000
    timeout client  60000
    timeout server  60000
    maxconn 320000
    errorfile 400 /etc/haproxy/errors/400.http
    errorfile 403 /etc/haproxy/errors/403.http
    errorfile 408 /etc/haproxy/errors/408.http
    errorfile 500 /etc/haproxy/errors/500.http
    errorfile 502 /etc/haproxy/errors/502.http
    errorfile 503 /etc/haproxy/errors/503.http
    errorfile 504 /etc/haproxy/errors/504.http

frontend haproxynode
    bind *:80
    bind *:443 ssl crt /etc/ssl/<redacted>/<redacted>.com.pem

    # Redirect if HTTPS is *not* used
    redirect scheme https code 301 if { hdr(Host) -i <redacted> -i <redacted> } !{ ssl_fc }
    
    acl health-endpoint-acl path_beg /haproxy/health
    use_backend http_200 if health-endpoint-acl

    acl repBack-acl hdr(host) -i <redacted> -i <redacted>:80
    use_backend repBack if repBack-acl

    default_backend backendnodes

    # both request and response headers must be captured for filebeat to work
    capture request header User-Agent len 256
    capture request header Host len 256
    capture response header CV len 15

backend backendnodes
    balance roundrobin
    option forwardfor
    http-request set-header X-Forwarded-Port %[dst_port]
    http-request add-header X-Forwarded-Proto https if { ssl_fc }
    http-response del-header server
    option httpchk GET /health HTTP/1.1\r\nHost:<redacted>
    timeout check 2000
    option redispatch
    retry-on empty-response
    server www-0 <redacted> check maxconn 8192
    server www-1 <redacted> check maxconn 8192


backend repBack
    balance source
    option forwardfor
    http-request set-header X-Forwarded-Port %[dst_port]
    http-request add-header X-Forwarded-Proto https if { ssl_fc }
    http-response del-header server
    option httpchk GET /repBack/version HTTP/1.1\r\nHost:<redacted>
    server repBack <redacted> check maxconn 1024

backend http_200
    http-request use-service lua.200-response

listen stats
    bind :32700
    stats enable
    stats uri /
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.4.11-1ppa1~focal 2022/01/07 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.11.html
Running on: Linux 5.4.0-66-generic #74-Ubuntu SMP Wed Jan 27 22:54:38 UTC 2021 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -O2 -fdebug-prefix-map=/build/haproxy-Z8Nhc7/haproxy-2.4.11=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_SYSTEMD=1 USE_PROMEX=1
  DEBUG   =

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=16).
Built with OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
Running on OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.34 2019-11-21
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 9.3.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [SPOE] spoe
        [CACHE] cache
        [FCGI] fcgi-app
        [COMP] compression
        [TRACE] trace
```

### Last Outputs and Backtraces

```plain
Haproxy did not crash so i don't think this is relevant
```

### Additional Information

The endpoint causing the issue in my config is the endpoint called "repBack"

**Comment by trebonius0** [comment]
Edit: 
It seems the issue is also affecting http 1.1 and not only http 1.0 as I initially though

**Comment by trebonius0** [comment]
Edit2: haproxy 2.2 doesn't have the issue (I switched to that branch)

**Comment by lukastribus** [comment]
Does this happen with both HTTP and HTTPS, or are you just using HTTPS here?
If this happens with HTTPS, could you try with HTTP and `option splice-response` in the backend?

If this is a regression in 2.4.11, the following commits sound related @capflam

```
BUG/MEDIUM: http-ana: Preserve response's FLT_END analyser on L7 retry
BUG/MINOR: mux-h1: Fix splicing for messages with unknown length
BUG/MEDIUM: mux-h1: Fix splicing by properly detecting end of message
```

**Comment by trebonius0** [comment]
It happens with both http & https (the initial issue I saw was with http only and then I realized it was far more important)

**Comment by capflam** [comment]
@lukastribus has right, it is most probably one of the H1 mux bugfix. With a quick test, I'm unable to reproduce the bug. Do you know if Apache announces a content-length ? Could you also confirm or inform it is a 2.4.11 regression ?

**Comment by xorander00** [comment]
Just chiming in to say that I'm seeing the same behavior on my server. I haven't checked http yet, but it's happening on https and I've confirmed that it's not a backend issue. If I get any more info, I'll drop it here if it's helpful.
  
- FreeBSD 13.0-STABLE 1300522
- HAproxy 2.4.11-1e183b8 2022/01/07

**Comment by trebonius0** [comment]
@capflam Very probably a 2.4.11 regression, as I did the upgrade yesterday evening and got complaints this morning. 
Yeah my Apache returns a content-length

**Comment by idl0r** [comment]
I tested 2.4.10 some minutes ago, with some additional patches though, but it doesn't seem to be affected, re #1511.
So I'd tend to agree that only 2.4.11 is affected.

**Comment by capflam** [comment]
Is it possible to share response headers from Apache and HAproxy ? If it is possible, you may try to enable traces on the H1 mux via the following command:
```
(echo "trace h1 sink buf0; trace h1 level developer; trace h1 verbosity complete; trace h1 start now; show events buf0 -w -n" ; cat ) | socat - {HA-SOCK-PATH}
```

It is pretty verbose, but this may help to figure out what happens.

**Comment by roedie** [comment]
I think I'm hitting the same issue but not with downloading, but with uploading of big-ish (>70MB) TIFF files.

This is shown in my log:
```
Jan 10 12:09:45 [redacted] haproxy[21823]: 2001:67c:[redacted] [10/Jan/2022:12:09:20.543] [redacted]~ [redacted]/[redacted] 11198/0/0/-1/25137 500 1991 - - IH-- 957/282/0/0/0 0/0 {[redacted].[redacted].com|Mozilla/5.0 (Mac|80349066|https://[redacted].[redacted].com/upload} {} “POST https://[redacted].[redacted].com/upload/process?projectId=3431&setId=149 HTTP/2.0”
```

I'm not using splicing, but the frontend is talking http/2 and the backend http/1.1.

**Comment by mildis** [comment]
Issue here too with session state `ID--` (like #1511)
Rolling back to 2.4.10 correct this behaviour.

**Comment by trebonius0** [comment]
@capflam It's weird, for some reason i cannot reproduce the issue anymore, in prod or in a vm

The apache headers are the following (for a simple http/1.1 request)
```
Date: Mon, 10 Jan 2022 14:46:03 GMT
Server: Apache
Last-Modified: Mon, 10 Jan 2022 01:13:37 GMT
ETag: "29eac-5d5300bd66cf6"
Accept-Ranges: bytes
Content-Length: 171692
Cache-Control: max-age=2592000, public
Content-Type: image/png

```

**Comment by capflam** [comment]
Thanks @trebonius0. @roedie, @mildis, could you share your config please ? Anyone is able to reproduce the issue on a dedicated platform to be able to enable traces on the H1 multiplexer ?

**Comment by xorander00** [comment]
@capflam Mine is dedicated. I have to hop on a a call here shortly though. If I'm not able to get to it before, I'll do it right after unless you're able to reproduce it before then.

**Comment by trebonius0** [comment]
@capflam i shared my config in the initial post of this thread :)

**Comment by capflam** [comment]
@xorander00 , many thanks. I'm still trying to reproduce it with no luck for now. @trebonius0, I know. I asked for the config of the other reporters :)

**Comment by xorander00** [comment]
@capflam Just dumped it, and yes it is very verbose lol. I enabled it via the sock & immediately loaded a page. What's funny is that the page loaded fine with the trace enabled (which I did dump to a file). I stopped the trace and reloaded the page and it went back to returning a response size mismatch.

**Comment by xorander00** [comment]
@capflam Which part of the log should I snip? I do need to do a quick review & scrub it of anything that could be sensitive, and I don't have the time to do that for the 3mb log it produced in the span of a few seconds lol.  
  
I just did a search for the hostname I accessed. Will that section + a bit of context before and after suffice?

**Comment by xorander00** [comment]
@capflam [trace.log](https://github.com/haproxy/haproxy/files/7839896/trace.log)

**Comment by capflam** [comment]
If there is no error in the traces, it is useless to share it. It may be easier to just log errors in the H1 traces. To do so, you should set the trace level to `error`:
```
(echo "trace h1 sink buf0; trace h1 level error; trace h1 verbosity complete; trace h1 start now; show events buf0 -w -n" ; cat ) | socat - {HA-SOCK-PATH}
```

**Comment by capflam** [comment]
@xorander00, I will take a look to your log, many thanks !

**Comment by xorander00** [comment]
@capflam That's the log with level = developer. Let me dump another one with level = error.

**Comment by capflam** [comment]
Thanks !

**Comment by mildis** [comment]
haproxy running at the edge, TLS-termination : 
```
global
	log /dev/log	local0 info
	log /dev/log	local1 notice
	chroot /var/lib/haproxy
	stats socket /run/haproxy/admin.sock mode 660 level admin expose-fd listeners
	stats timeout 30s
	user haproxy
	group haproxy
	daemon

	# Default SSL material locations
	ca-base /etc/ssl/certs
	crt-base /etc/ssl/private

	ssl-default-bind-ciphers TLSv1.3:TLSv1.2:!RSA:!SHA1:!SHA256:!SHA384:!aNULL:!eNULL
	ssl-default-bind-options ssl-min-ver TLSv1.2

	tune.ssl.default-dh-param 2048


defaults
	log	global
	mode	http
	option	httplog
	option	dontlognull
        timeout connect 5000
        timeout client  50000
        timeout server  50000
	errorfile 400 /etc/haproxy/errors/400.http
	errorfile 403 /etc/haproxy/errors/403.http
	errorfile 408 /etc/haproxy/errors/408.http
	errorfile 500 /etc/haproxy/errors/500.http
	errorfile 502 /etc/haproxy/errors/502.http
	errorfile 503 /etc/haproxy/errors/503.http
	errorfile 504 /etc/haproxy/errors/504.http

	option forwardfor       except 127.0.0.0/8
	option                  redispatch
#	option http-server-close
	option http-keep-alive
#	no option http-use-htx
	option tcp-smart-accept

listen adminstats
    bind 127.0.0.1:8080
    mode http
    balance
    timeout client  5000
    timeout connect 4000
    timeout server 30000
    stats enable
    stats uri /haproxy_stats       
    stats realm HAProxy\ Statistics
    stats auth john:doe

frontend ft-https
    bind *:443 name https tfo ssl alpn http/1.1 crt /etc/ssl/private/certificate-2022.crt

    maxconn 200

#    capture request  header Host len 50
#    capture response header Location len 50

    http-request track-sc0 src table per_ipv4_rates

    http-request set-header X-Forwarded-Proto https
    http-request set-header X-Forwarded-Port 443
    http-request set-header X-Forwarded-Host %[req.hdr(host),lower]


    http-response set-header Strict-Transport-Security max-age=31536000;\ includeSubDomains;
    http-response del-header Server
    http-response del-header X-AspNet-Version
    http-response del-header X-Powered-By

    acl secured_cookie res.hdr(Set-Cookie),lower -m sub secure
    http-response replace-header Set-Cookie (.*) \1;\ Secure unless secured_cookie

   use_backend %[req.hdr(host),lower]

backend per_ipv4_rates
 stick-table type ip size 1m expire 10m store http_req_rate(10s)

backend myserver.url
  timeout server 30m
  server server1 192.168.1.2:8082 send-proxy-v2 check
```


haproxy running within the backend server : 
```
global
	log /dev/log	local0 info
	log /dev/log	local1 notice
	chroot /var/lib/haproxy
	stats socket /run/haproxy-admin.sock mode 660 level admin
	stats timeout 30s
	user haproxy
	group haproxy
	daemon

	# Default SSL material locations
	ca-base /etc/ssl/certs
	crt-base /etc/pki/tls/private

	# Default ciphers to use on SSL-enabled listening sockets.
	# For more information, see ciphers(1SSL). This list is from:
	#  https://hynek.me/articles/hardening-your-web-servers-ssl-ciphers/
#	ssl-default-bind-ciphers ECDH+AESGCM:DH+AESGCM:ECDH+AES256:DH+AES256:ECDH+AES128:DH+AES:ECDH+3DES:DH+3DES:RSA+AESGCM:RSA+AES:RSA+3DES:!aNULL:!MD5:!DSS
#	tune.ssl.default-dh-param 1024
	ssl-default-bind-ciphers EECDH+AES:+AES128:+AES256:+SHA:DHE-RSA-AES256-SHA:DHE-RSA-AES128-SHA
	ssl-default-bind-options no-sslv3


defaults
	log	global
	mode	http
	option	httplog
	option	dontlognull
        timeout connect 5s
        timeout client  5m
        timeout server  30m
	errorfile 400 /etc/haproxy/errors/400.http
	errorfile 403 /etc/haproxy/errors/403.http
	errorfile 408 /etc/haproxy/errors/408.http
	errorfile 500 /etc/haproxy/errors/500.http
	errorfile 502 /etc/haproxy/errors/502.http
	errorfile 503 /etc/haproxy/errors/503.http
	errorfile 504 /etc/haproxy/errors/504.http

	option forwardfor       except 127.0.0.0/8
	option                  redispatch
#	option http-server-close
#	option http-keep-alive

listen adminstats
    bind 127.0.0.1:8081
    mode http
    balance
    timeout client  5000
    timeout connect 4000
    timeout server 30000
    stats enable
    stats uri /haproxy_stats       
    stats realm HAProxy\ Statistics
    stats auth john:doe

frontend main
    bind 192.168.1.2:8082 accept-proxy

#    capture request  header Host len 50
#    capture response header Location len 50

    http-request deny unless { hdr_dom(host) -i myserver.url }

    http-request set-header X-Forwarded-Proto https
    http-request set-header X-Forwarded-Port 443
    http-request set-header X-Forwarded-Host %[req.hdr(host),lower]

    http-response set-header Strict-Transport-Security max-age=31536000;\ includeSubDomains;

    acl secured_cookie res.hdr(Set-Cookie),lower -m sub secure
    http-response replace-header Set-Cookie (.*) \1;\ Secure unless secured_cookie

    acl to-root path -i /
    acl to-prod	path_beg /prod

    redirect code 301 location https://myserver.url/prod/ if to-root

    use_backend bck-prod if to-prod

backend bck-prod
    acl hdr_location res.hdr(Location) -m found
    http-response replace-header Location https?(://.*) https\1 if hdr_location
    server sx-prod 127.0.0.1:8080 check
```

**Comment by xorander00** [comment]
He's looking for runtime trace info as it happens, not configs.

**Comment by idl0r** [comment]
> He's looking for runtime trace info as it happens, not configs.

I think both is fine, as it is difficult to reproduce at the moment.

**Comment by xorander00** [comment]
@capflam Looks like we have some paydirt.  Here you go...  
  
```
<0>2022-01-10T10:30:32.033465-05:00 [00|h1|0|mux_h1.c:2180] too much payload, more than announced : [F] [MSG_DONE, MSG_DATA] - req=(.fl=0x00001550 .curr_len=0 .body_len=0)  res=(.fl=0x00001515 .curr_len=4352 .body_len=195976) - h1c=0x8058c9d40(0x00002200) conn=0x8058ee300(0x80003300) h1s=0x8059b6ec0(0x00104010) cs=0x8011fd880(0x00004000)
<0>2022-01-10T10:30:32.033481-05:00 [00|h1|0|mux_h1.c:2314] processing output error, set error on h1c/h1s : [F] [MSG_DONE, MSG_DATA] - req=(.fl=0x00001550 .curr_len=0 .body_len=0)  res=(.fl=0x00001515 .curr_len=4352 .body_len=195976) - h1c=0x8058c9d40(0x00002a00) conn=0x8058ee300(0x80003300) h1s=0x8059b6ec0(0x00105010) cs=0x8011fd880(0x00004000)
<0>2022-01-10T10:30:32.033484-05:00 [00|h1|0|mux_h1.c:3504] reporting error to the app-layer stream : [F] [MSG_DONE, MSG_DATA] - req=(.fl=0x00001550 .curr_len=0 .body_len=0)  res=(.fl=0x00001515 .curr_len=4352 .body_len=195976) - h1c=0x8058c9d40(0x00102a00) conn=0x8058ee300(0x80003300) h1s=0x8059b6ec0(0x00105010) cs=0x8011fd880(0x00004100)
<0>2022-01-10T10:30:32.033541-05:00 [00|h1|0|/mux_h1.c:756] h1s on error, set error on h1c : [F] [MSG_DONE, MSG_DATA] - req=(.fl=0x00001550 .curr_len=0 .body_len=0)  res=(.fl=0x00001515 .curr_len=4352 .body_len=195976) - h1c=0x8058c9d40(0x00005800) conn=0x8058ee300(0x80003300) h1s=0x8059b6ec0(0x00105010)
<0>2022-01-10T10:30:32.061049-05:00 [00|h1|0|mux_h1.c:1828] too much payload, more than announced : [F] [MSG_DONE, MSG_DATA] - req=(.fl=0x00001550 .curr_len=0 .body_len=0)  res=(.fl=0x00001515 .curr_len=2823 .body_len=1517488) - h1c=0x805936f40(0x00102200) conn=0x8058ee240(0x80003300) h1s=0x805921400(0x00104010) cs=0x801056800(0x00004000)
<0>2022-01-10T10:30:32.061059-05:00 [00|h1|0|mux_h1.c:2314] processing output error, set error on h1c/h1s : [F] [MSG_DONE, MSG_DATA] - req=(.fl=0x00001550 .curr_len=0 .body_len=0)  res=(.fl=0x00001515 .curr_len=2823 .body_len=1517488) - h1c=0x805936f40(0x00102a00) conn=0x8058ee240(0x80003300) h1s=0x805921400(0x00105010) cs=0x801056800(0x00004000)
<0>2022-01-10T10:30:32.061062-05:00 [00|h1|0|mux_h1.c:3504] reporting error to the app-layer stream : [F] [MSG_DONE, MSG_DATA] - req=(.fl=0x00001550 .curr_len=0 .body_len=0)  res=(.fl=0x00001515 .curr_len=2823 .body_len=1517488) - h1c=0x805936f40(0x00102a00) conn=0x8058ee240(0x80003300) h1s=0x805921400(0x00105010) cs=0x801056800(0x00004100)
<0>2022-01-10T10:30:32.061154-05:00 [00|h1|0|/mux_h1.c:756] h1s on error, set error on h1c : [F] [MSG_DONE, MSG_DATA] - req=(.fl=0x00001550 .curr_len=0 .body_len=0)  res=(.fl=0x00001515 .curr_len=2823 .body_len=1517488) - h1c=0x805936f40(0x00005800) conn=0x8058ee240(0x80083300) h1s=0x805921400(0x00105010)
<0>2022-01-10T10:30:35.353737-05:00 [00|h1|0|mux_h1.c:1828] too much payload, more than announced : [F] [MSG_DONE, MSG_DATA] - req=(.fl=0x00001550 .curr_len=0 .body_len=0)  res=(.fl=0x00001515 .curr_len=13894 .body_len=195976) - h1c=0x8058ee300(0x00102200) conn=0x8058f0940(0x80003300) h1s=0x8058ee780(0x00104010) cs=0x8011fdd40(0x00004000)
<0>2022-01-10T10:30:35.353766-05:00 [00|h1|0|mux_h1.c:2314] processing output error, set error on h1c/h1s : [F] [MSG_DONE, MSG_DATA] - req=(.fl=0x00001550 .curr_len=0 .body_len=0)  res=(.fl=0x00001515 .curr_len=13894 .body_len=195976) - h1c=0x8058ee300(0x00102a00) conn=0x8058f0940(0x80003300) h1s=0x8058ee780(0x00105010) cs=0x8011fdd40(0x00004000)
<0>2022-01-10T10:30:35.353770-05:00 [00|h1|0|mux_h1.c:3504] reporting error to the app-layer stream : [F] [MSG_DONE, MSG_DATA] - req=(.fl=0x00001550 .curr_len=0 .body_len=0)  res=(.fl=0x00001515 .curr_len=13894 .body_len=195976) - h1c=0x8058ee300(0x00102a00) conn=0x8058f0940(0x80003300) h1s=0x8058ee780(0x00105010) cs=0x8011fdd40(0x00004100)
<0>2022-01-10T10:30:35.353839-05:00 [00|h1|0|/mux_h1.c:756] h1s on error, set error on h1c : [F] [MSG_DONE, MSG_DATA] - req=(.fl=0x00001550 .curr_len=0 .body_len=0)  res=(.fl=0x00001515 .curr_len=13894 .body_len=195976) - h1c=0x8058ee300(0x00005800) conn=0x8058f0940(0x80083300) h1s=0x8058ee780(0x00105010)
<0>2022-01-10T10:30:35.378172-05:00 [00|h1|0|mux_h1.c:1828] too much payload, more than announced : [F] [MSG_DONE, MSG_DATA] - req=(.fl=0x00001550 .curr_len=0 .body_len=0)  res=(.fl=0x00001515 .curr_len=6019 .body_len=1517488) - h1c=0x8058ee240(0x00102200) conn=0x8058ef5c0(0x80003300) h1s=0x8058eee40(0x00104010) cs=0x8011fd600(0x00004000)
<0>2022-01-10T10:30:35.378181-05:00 [00|h1|0|mux_h1.c:2314] processing output error, set error on h1c/h1s : [F] [MSG_DONE, MSG_DATA] - req=(.fl=0x00001550 .curr_len=0 .body_len=0)  res=(.fl=0x00001515 .curr_len=6019 .body_len=1517488) - h1c=0x8058ee240(0x00102a00) conn=0x8058ef5c0(0x80003300) h1s=0x8058eee40(0x00105010) cs=0x8011fd600(0x00004000)
<0>2022-01-10T10:30:35.378185-05:00 [00|h1|0|mux_h1.c:3504] reporting error to the app-layer stream : [F] [MSG_DONE, MSG_DATA] - req=(.fl=0x00001550 .curr_len=0 .body_len=0)  res=(.fl=0x00001515 .curr_len=6019 .body_len=1517488) - h1c=0x8058ee240(0x00102a00) conn=0x8058ef5c0(0x80003300) h1s=0x8058eee40(0x00105010) cs=0x8011fd600(0x00004100)
<0>2022-01-10T10:30:35.378277-05:00 [00|h1|0|/mux_h1.c:756] h1s on error, set error on h1c : [F] [MSG_DONE, MSG_DATA] - req=(.fl=0x00001550 .curr_len=0 .body_len=0)  res=(.fl=0x00001515 .curr_len=6019 .body_len=1517488) - h1c=0x8058ee240(0x00005800) conn=0x8058ef5c0(0x80083300) h1s=0x8058eee40(0x00105010)
```

**Comment by capflam** [comment]
Ok, so, that confirms 5f60d18a2c is most probably the faulty commit. But I'm still don't understand the reason.

**Comment by xorander00** [comment]
@capflam I take it you were able to reproduce it?

**Comment by capflam** [comment]
not yet

**Comment by capflam** [comment]
I guess I found the issue. At least for #1511 (about internal error reports). I'm not convinced it is related to this one. here is the patch:

[0001-BUG-MAJOR-mux-h1-Don-t-decrement-.curr_len-for-unsen.patch.txt](https://github.com/haproxy/haproxy/files/7840388/0001-BUG-MAJOR-mux-h1-Don-t-decrement-.curr_len-for-unsen.patch.txt)

@xorander00, could you try it ? @trebonius0, you may also try it, just in case.

**Comment by capflam** [comment]
@trebonius0, if your client presets the file size with the announced content length before dumping the response payload, because an internal error is reported and the transfer is interrupted, this may explain why the file size is correct but the file is padded with 0x00.

**Comment by trebonius0** [comment]
That may be why indeed as it's exactly what my code behind my apache server is doing

**Comment by xorander00** [comment]
@capflam Will give the patch a shot here shortly. I build it on FreeBSD via the port, so I'll have to merge the patch into it there.

**Comment by xorander00** [comment]
@capflam Yup, that patch appears to have fix it (for me at least anyway).

**Comment by capflam** [comment]
Thanks @xorander00, I'll push the patch then.

**Comment by vincentbernat** [comment]
Do you plan to release a 2.4.12 really soon or should we push the patch to users on top of 2.4.11?

**Comment by lukastribus** [comment]
I suggest releasing `2.4.12` with the fix.

**Comment by roedie** [comment]
Would this also fix the upload problem? Or is that a different bug?

I can try the patch later today to test that...

**Comment by capflam** [comment]
@roedie, it is indeed the same issue. And a 2.4.12 will be released very soon.

**Comment by idl0r** [comment]
> @roedie, it is indeed the same issue. And a 2.4.12 will be released very soon.

Will there be other fixes as well, in .12? I'm building 2.4.11 with your patch right now and going to deploy it on 2 production LBs.

**Comment by capflam** [comment]
f82afbb9c will be backported too.

**Comment by capflam** [comment]
Backported as far as 2.4. I will emit a 2.4.12. Thanks guys !

**Comment by capflam** [comment]
FYI, I'll also emit a 2.5.1.

**Comment by idl0r** [comment]
Perfect, thanks, Christopher!

**Comment by idl0r** [comment]
FYI: Right now, there's no more errors logged on the patched 2.4.11 LBs. Prior to the patch, the logs contained some errors immediately and continuously.
So at a first glance, it looks good!

**Comment by capflam** [comment]
Thanks for your feedback @idl0r !
